examples/vizdoc/html/heatmap.html

```diff
@@ -0,0 +1,193 @@
+<!DOCTYPE html>
+<svg width="960" height="500" stroke="#fff" stroke-width="0.5"></svg>
+<p>
+    <label for="circ1x" 
+       style="display: inline-block; width: 240px; text-align: right">
+       circle 1 x-coord = <span id="circ1x-value">…</span>
+    </label>
+    <input type="range" min="0" max="960" id="circ1x">
+    <label for="circ1y" 
+       style="display: inline-block; width: 240px; text-align: right">
+       circle 1 y-coord = <span id="circ1y-value">…</span>
+    </label>
+    <input type="range" min="0" max="500" id="circ1y">
+  </p>
+  <p>
+    <label for="circ2x" 
+       style="display: inline-block; width: 240px; text-align: right">
+       circle 2 x-coord = <span id="circ2x-value">…</span>
+    </label>
+    <input type="range" min="0" max="960" id="circ2x">
+    <label for="circ2y" 
+       style="display: inline-block; width: 240px; text-align: right">
+       circle 2 y-coord = <span id="circ2y-value">…</span>
+    </label>
+    <input type="range" min="0" max="500" id="circ2y">
+  </p>
+<script src="https://d3js.org/d3.v4.min.js"></script>
+<script src="https://d3js.org/d3-contour.v1.min.js"></script>
+<script src="https://d3js.org/d3-scale-chromatic.v1.min.js"></script>
+<script>
+
+// most of this code is taken from https://blockbuilder.org/mbostock/f48ff9c1af4d637c9a518727f5fdfef5
+// currently I want to illustrate the sameCenter objective. However, this base code could be repurposed to illustrate many functions
+// if I can get it to work, that is.
+
+// INITIALIZATION
+
+// TODO- translate d3 coords to penrose coords
+let circ1 = [90, 100] // circles we will attempt to sameCenter, initially fixed locations
+let circ2 = [500, 80]
+
+// Populate a grid of n×m values - these correspond to energy outputs
+var n = 240, m = 125, values = new Array(n * m);
+
+// Initialize each cell with energy of obj function
+function calcval(){
+    for (var j = 0, k = 0; j < m; ++j) {
+    for (var i = 0; i < n; ++i, ++k) {
+        values[k] = constraint(i, j);
+    }
+    }
+}
+
+// Sum the distances between a cell and the two circles - is this the right function
```

Comment:
So this is almost the right function, but not quite :) `sameCenter` is a function with four DOF: x1, y1, x2, y2, defined as ||c2-c1||^2 (where c1 = [x1, y2], etc). 

When you draw the energy function, it's only possible to plot two DOF in 2D, just an x and y.

So, really you should fix one of the points `c1`, and calculate sameCenter over all cells, using the distance from the moveable point `c2`. So the energy at cell `p := (x,y)` is `distsq(p, c1)`. The result is basically a "bowl" at `c1`, with no moveable point `c2`, because the energy function is taken over all points `c2`. But `c1` can move, and when it does, the bowl should move with it. Does that make sense? I would change the math here (and the sliders and README explanation) to reflect this. (I also recommend doing math in terms of vector operations, like norm, rather than coordinate-wise operations. It's just way more readable.)

Now, whether that is a good way to visualize the function is another question. Probably a better way is to show the gradient of sameCenter as two arrows `[df/dx1, df/fy1]` and `[df/dx2, df/dy2]`, which shows "everything" rather than the limited slice from the heatmap. But we can leave that for another time.

Reply:
Tried fixing it in 0e5d3a803d6d199268633fdf5d6c1d38a92acaea @hypotext , still might not be right! Also, how would you recommend doing math with vectors? I was hesitant to use TFJS since it seemed like it was being factored out...

Reply:
Yes, this looks right to me. Don't worry about the vector math (especially since you're making the visualization in plain JS), though for future reference I'd just define some plain JS helper functions on lists, or use lodash.